### PR TITLE
Support function calling in Ollama API

### PR DESCRIPTION
**Pull request by ThomasVitale**
* Update Ollama API types with the latest definitions for Generation and Chat Completion.
* Include support for function calling at the API level + integration tests.
* Refine integration tests to use latest Ollama version and parametrized models.

A second PR will raise the function calling support at the ChatModel level.

Fixes gh-720

### CODE REVIEW

File: models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java [hunk 1]
```diff
@@ -122,59 +125,43 @@ public OllamaApi(String baseUrl, RestClient.Builder restClientBuilder) {
...
+	 * @param images A list of base64-encoded images accompanying this request, for multimodal models such as "llava".
+	 * @param options Model-specific options. For example, "temperature" can be set through this field, if the model supports it.
+	 *
+	 * @see <a href=
+	 * "https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-completion">Generation API</a>
+	 * @see <a href="https://github.com/ollama/ollama/blob/main/api/types.go">Ollama
+	 * Types</a>
 	 */
 	@JsonInclude(Include.NON_NULL)
 	public record GenerateRequest(
```
Comment (ThomasVitale) [inline models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java:147]: It's worth considering whether we want to keep maintaining the APIs for the /generate endpoint which is not used by any Spring AI abstraction.

For now, I changed it so that it's fully in sync with the latest Ollama version, including the order of the fields to help future maintenance.

File: models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java [hunk 2]
```diff
@@ -122,59 +125,43 @@ public OllamaApi(String baseUrl, RestClient.Builder restClientBuilder) {
...
 			@JsonProperty("stream") Boolean stream,
 			@JsonProperty("raw") Boolean raw,
+			@JsonProperty("format") String format,
+			@JsonProperty("keep_alive") String keepAlive,
 			@JsonProperty("images") List<String> images,
-			@JsonProperty("keep_alive") String keepAlive) {
+			@JsonProperty("options") Map<String, Object> options
+	) {
 
-		/**
```
Comment (ThomasVitale) [inline models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java:159]: Removed these overloaded constructors since the `Builder` object can be used to construct a new object in a safer and more maintainable way (in particular, in order to prevent bugs related to misplaced "null" values).

File: models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java [hunk 3]
```diff
@@ -374,19 +374,44 @@ public enum Role {
...
+		/**
+		 * The function definition for a tool call.
+		 *
+		 * @param name The name of the function.
+		 * @param arguments The arguments that the model expects you to pass to the function.
+		 */
+		@JsonInclude(Include.NON_NULL)
+		public record ToolCallFunction(
+				@JsonProperty("name") String name,
+				@JsonProperty("arguments") Map<String,Object> arguments) {
```
Comment (ThomasVitale) [inline models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java:403]: This is very similar to the OpenAI API. The main difference is that OpenAI returns arguments as a JSON String, whereas Ollama returns them in a Map.

Comment (ThomasVitale) [reply models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaApi.java:403]: The other difference is that Ollama doesn't currently handle IDs for ToolCalls. Perhaps they will add that in the future.

File: models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaOptions.java [hunk 4]
```diff
@@ -64,12 +66,6 @@ public class OllamaOptions implements ChatOptions, EmbeddingOptions {
 	 */
 	@JsonProperty("num_batch") private Integer numBatch;
 
-	/**
-	 * The number of GQA groups in the transformer layer. Required for some models,
-	 * for example it is 8 for llama2:70b.
-	 */
-	@JsonProperty("num_gqa") private Integer numGQA;
```
Comment (ThomasVitale) [inline models/spring-ai-ollama/src/main/java/org/springframework/ai/ollama/api/OllamaOptions.java:71]: This option is not supported anymore.

File: models/spring-ai-ollama/src/test/java/org/springframework/ai/ollama/OllamaChatModelIT.java [hunk 5]
```diff
@@ -58,18 +59,18 @@
...
 class OllamaChatModelIT {
 
-	private static String MODEL = "mistral";
+	private static final String MODEL = OllamaModel.MISTRAL.getName();
 
 	private static final Log logger = LogFactory.getLog(OllamaChatModelIT.class);
 
 	@Container
-	static OllamaContainer ollamaContainer = new OllamaContainer("ollama/ollama:0.1.32");
+	static OllamaContainer ollamaContainer = new OllamaContainer("ollama/ollama:0.2.8");
```
Comment (ThomasVitale) [inline models/spring-ai-ollama/src/test/java/org/springframework/ai/ollama/OllamaChatModelIT.java:67]: The Ollama integration test setup could use some cleanup and refactoring, but let's keep that for another PR to keep the scope focused here.

**Comment by tzolov** [comment]
This should be covered by the d53876a6244a71efb471f3934c4e1dcf08268b2a and b9593fe6f0500bc6e040f5ab6fff0251e4a950c0

@ThomasVitale feel free to re-open it if I've missed something.
